File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomDevicesFragment.java
package com.launchkey.android.authenticator.demo.ui.fragment;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.snackbar.Snackbar;
import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.demo.databinding.DemoFragmentDevicesBinding;
import com.launchkey.android.authenticator.demo.ui.adapter.DemoDevicesAdapter;
import com.launchkey.android.authenticator.demo.util.Utils;
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorManager;
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device;
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.GetDevicesEventCallback;
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback;
import com.launchkey.android.authenticator.sdk.core.util.CompositeDisposable;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class CustomDevicesFragment extends BaseDemoFragment<DemoFragmentDevicesBinding>
        implements AdapterView.OnItemClickListener {

    private final @NonNull CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final AuthenticatorManager deviceManager = AuthenticatorManager.instance;
    private List<Device> devices = new ArrayList<>();
    private DemoDevicesAdapter adapter;
    private final GetDevicesEventCallback getDevicesEventCallback = new GetDevicesEventCallback() {
        @Override
        public void onSuccess(@NonNull List<Device> devices) {
            CustomDevicesFragment.this.devices.clear();
            CustomDevicesFragment.this.devices.addAll(devices);
            adapter.notifyDataSetChanged();
        }

        @Override
        public void onFailure(@NonNull Exception e) {
            Utils.simpleSnackbarForBaseError(binding.demoFragmentDevicesList, e);
        }
    };

    public CustomDevicesFragment() {
        super(R.layout.demo_fragment_devices);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = DemoFragmentDevicesBinding.bind(view);
        adapter = new DemoDevicesAdapter(getActivity(), devices, this);
        binding.demoFragmentDevicesList.setAdapter(adapter);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState == null) {
            compositeDisposable.add(deviceManager.getDevices(getDevicesEventCallback));
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        compositeDisposable.clear();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        if (AuthenticatorManager.instance.isDeviceLinked()) {
            Device deviceToUnlink = adapter.getItem(position);
            compositeDisposable.add(deviceManager.unlinkDevice(deviceToUnlink, new UnlinkDeviceEventCallback() {
                @Override
                public void onSuccess(final @NonNull Device device) {
                    String message = String.format(Locale.getDefault(), "Device \"%s\" unlinked", device.getName());
                    Snackbar.make(binding.demoFragmentDevicesList, message, Snackbar.LENGTH_INDEFINITE)
                            .setAction("Return", new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    Utils.finish(CustomDevicesFragment.this);
                                }
                            })
                            .show();
                    if (!device.isCurrent()) {
                        deviceManager.getDevices(getDevicesEventCallback);
                    }
                }

                @Override
                public void onFailure(final @NonNull Exception e) {
                    Utils.simpleSnackbarForBaseError(binding.demoFragmentDevicesList, e);
                }
            }));
        }
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/fragment/CustomDevicesFragment.kt
package com.launchkey.android.authenticator.demo.ui.fragment

import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import com.google.android.material.snackbar.Snackbar
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.demo.databinding.DemoFragmentDevicesBinding
import com.launchkey.android.authenticator.demo.ui.adapter.DemoDevicesAdapter
import com.launchkey.android.authenticator.demo.util.Utils.finish
import com.launchkey.android.authenticator.demo.util.Utils.simpleSnackbarForBaseError
import com.launchkey.android.authenticator.sdk.core.authentication_management.AuthenticatorManager
import com.launchkey.android.authenticator.sdk.core.authentication_management.Device
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.GetDevicesEventCallback
import com.launchkey.android.authenticator.sdk.core.authentication_management.event_callback.UnlinkDeviceEventCallback
import com.launchkey.android.authenticator.sdk.core.util.CompositeDisposable
import java.util.*

class CustomDevicesFragment : BaseDemoFragment<DemoFragmentDevicesBinding>(R.layout.demo_fragment_devices), OnItemClickListener {
    private val compositeDisposable = CompositeDisposable()
    private val deviceManager = AuthenticatorManager.instance
    private val devices: MutableList<Device> = ArrayList()
    private var adapter: DemoDevicesAdapter? = null
    private val getDevicesEventCallback: GetDevicesEventCallback = object : GetDevicesEventCallback() {
        override fun onSuccess(devices: List<Device>) {
            this@CustomDevicesFragment.devices.clear()
            this@CustomDevicesFragment.devices.addAll(devices)
            adapter!!.notifyDataSetChanged()
        }

        override fun onFailure(e: Exception) {
            simpleSnackbarForBaseError(binding!!.demoFragmentDevicesList, e)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding = DemoFragmentDevicesBinding.bind(view)
        adapter = DemoDevicesAdapter(activity!!, devices, this)
        binding!!.demoFragmentDevicesList.adapter = adapter
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) {
            compositeDisposable.add(deviceManager.getDevices(getDevicesEventCallback))
        }
    }

    override fun onPause() {
        super.onPause()
        compositeDisposable.clear()
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
        if (AuthenticatorManager.instance.isDeviceLinked) {
            val deviceToUnlink = adapter!!.getItem(position)
            compositeDisposable.add(deviceManager.unlinkDevice(deviceToUnlink, object : UnlinkDeviceEventCallback() {
                override fun onSuccess(device: Device) {
                    val message = String.format(Locale.getDefault(), "Device \"%s\" unlinked", device.name)
                    Snackbar.make(binding!!.demoFragmentDevicesList, message, Snackbar.LENGTH_INDEFINITE)
                            .setAction("Return") { finish(this@CustomDevicesFragment) }
                            .show()
                    if (!device.isCurrent) {
                        deviceManager.getDevices(getDevicesEventCallback)
                    }
                }

                override fun onFailure(e: Exception) {
                    simpleSnackbarForBaseError(binding!!.demoFragmentDevicesList, e)
                }
            }))
        }
    }
}